src/features/courses/course-controller.ts

```diff
@@ -2522,6 +2641,250 @@ class CourseController {
+        const question = await courseRepository.getQuestion({ id: questionId });
+        const topic = await courseRepository.getCourseTopic({ id: question.courseTopicContentId });
+        return topic.topicTypeId === 2;
+    }
+
+    /*
+    * Determine if a user (student) is allowed to view a question -- profs+ => always true
+    * If the question belongs to an assessment, check for a current version unless a versionId is provided
+    */
+    async userCanViewQuestionId(user: User, questionId: number, studentTopicAssessmentInfoId?: number): Promise<boolean> {
+        if (user.roleId === Role.PROFESSOR || user.roleId === Role.ADMIN) return true;
+        const question = await courseRepository.getQuestion({ id: questionId });
+        const topic = await question.getTopic();
+        if (topic.topicTypeId === 1) {
+            return (topic.startDate.toMoment().isBefore(moment()));
+        } else if (topic.topicTypeId === 2) {
+            let topicIsLive = false;
+            if (_.isNil(studentTopicAssessmentInfoId)) {
+                // specific version was not supplied - see if there's a live version for this question
+                const gradeInstance = await courseRepository.getCurrentInstanceForQuestion({questionId, userId: user.id});
+                topicIsLive = !_.isNil(gradeInstance); // if we got a current instance, then topic is live
+            } else {
+                const studentTopicInfo = await this.getStudentTopicAssessmentInfoById(studentTopicAssessmentInfoId);
+                topicIsLive = studentTopicInfo.isClosed === false && // isClosed might not be accurate when the assessment times out
+                    moment().isBetween(studentTopicInfo.startTime.toMoment(), studentTopicInfo.endTime.toMoment());
+            }
+            const topicInfo = await topic.getTopicAssessmentInfo();
+            if (topicIsLive) {
+                return true;
+            } else {
+                // strike that -- reverse it ;)
+                return !topicInfo.hideProblemsAfterFinish;
+            }
+        }
+        // we *should* never get here - but if we do, then the answer is NO
+        return false;
+    };
+
+    scoreAssessment = (results: SubmittedAssessmentResultContext[]): ScoreAssessmentResult => {
+        let totalScore = 0;
+        let bestVersionScore = 0;
+        let bestOverallVersion = 0;
+        let problemScores: { [key: string]: number } = {};
+
+        results.forEach((result: SubmittedAssessmentResultContext) => {
+            const { questionResponse, grade, instance, weight } = result;
+            totalScore += weight * questionResponse.problem_result.score;
+            bestVersionScore += weight * instance.scoreForBestVersion;
+            bestOverallVersion += weight * grade.bestScore;
+            problemScores = {[result.instance.problemNumber.toString(10)]: result.questionResponse.problem_result.score * result.weight};
```

Comment:
This isn't doing what your thinking your doing (I think new lines would actually help make it clearer).
You're setting `problemScores` equal to a new object with a single fields set, thus none of the previous scores were kept, and you override it again below so only the total score would be kept

What you want to do:
```typescript
problemScores[result.instance.problemNumber.toString(10)] = result.questionResponse.problem_result.score * result.weight
```

Notice that the braces are gone and the key is used in brackets on the reference side

Reply:
I was suspicious of this, but couldn't work out how to use a variable as a key name...